### PR TITLE
fix(react-query): Allow optional initialData in infiniteQueryOptions

### PR DESCRIPTION
**Pull request by sungpaks**
Issue fix : #8132  
There was a typescript error(but could be run with no problem) when using optional `initialData` in `infiniteQueryOptions`

### CODE REVIEW

File: packages/react-query/src/infiniteQueryOptions.ts [hunk 1]
```diff
@@ -42,6 +42,7 @@ export type DefinedInitialDataInfiniteOptions<
   initialData:
     | NonUndefinedGuard<InfiniteData<TQueryFnData, TPageParam>>
     | (() => NonUndefinedGuard<InfiniteData<TQueryFnData, TPageParam>>)
+    | undefined
```
Comment (TkDodo) [inline packages/react-query/src/infiniteQueryOptions.ts:45]: pretty sure this should go to `UndefinedInitialDataInfiniteOptions`, like we have for `queryOptions`. compare:

https://github.com/TanStack/query/blob/fdb8ce1c38bc30fec9e984218b2fdc7802aef695/packages/react-query/src/queryOptions.ts#L15

**Review by TkDodo — CHANGES_REQUESTED** [review]
can you please add a type test to `packages/react-query/src/__tests__/infiniteQueryOptions.test-d.tsx` that fails before and succeeds with your fix ?

File: packages/react-query/src/infiniteQueryOptions.ts (hunk 1 above)
Comment (sungpaks) [reply packages/react-query/src/infiniteQueryOptions.ts:45]: Thankyou,
I just saw that you mentioned it in the issue. I'll reflect this.

But one thing what I wonder is, what is difference between two of them : 
- `(() => NonUndefinedGuard<InfiniteData<TQueryFnData, TPageParam>>)`
- `InitialDataFunction<NonUndefinedGuard<InfiniteData<TQueryFnData, TPageParam>>>`

Comment (TkDodo) [reply packages/react-query/src/infiniteQueryOptions.ts:45]: `InitialDataFunction` has `undefined` baked into the return value:

```
type InitialDataFunction<T> = () => T | undefined
```

Comment (sungpaks) [reply packages/react-query/src/infiniteQueryOptions.ts]: Thank you for your kindness!

**Comment by sungpaks** [comment]
> can you please add a type test to `packages/react-query/src/__tests__/infiniteQueryOptions.test-d.tsx` that fails before and succeeds with your fix ?

I added a type test that failed before and succeeds with my fix.

**Review by TkDodo — APPROVED** [review]

File: packages/react-query/src/infiniteQueryOptions.ts [hunk 2]
```diff
@@ -20,7 +21,11 @@ export type UndefinedInitialDataInfiniteOptions<
   TQueryKey,
   TPageParam
 > & {
-  initialData?: undefined
+  initialData?:
+    | undefined
+    | InitialDataFunction<
+        NonUndefinedGuard<InfiniteData<TQueryFnData, TPageParam>>
```
Comment (jimmycallin) [inline packages/react-query/src/infiniteQueryOptions.ts:27]: should this have been something like

```ts
  initialData?:
    | undefined
    | InfiniteData<TQueryFnData, TPageParam>
    | InitialDataFunction<
        NonUndefinedGuard<InfiniteData<TQueryFnData, TPageParam>>
```

to allow passing initialData as object and not within a callback function?

**Comment by jimmycallin** [comment]
thanks for addressing my issue so quickly! i added a post-merge comment with a potential overlook. :)

File: packages/react-query/src/infiniteQueryOptions.ts (hunk 2 above)
Comment (TkDodo) [reply packages/react-query/src/infiniteQueryOptions.ts:27]: we can pass initialData as an object via `DefinedInitialDataInfiniteOptions`:

https://github.com/TanStack/query/blob/febbde676d0888d8dd4868177868a3316686cb58/packages/react-query/src/infiniteQueryOptions.ts#L41-L45

Comment (jimmycallin) [reply packages/react-query/src/infiniteQueryOptions.ts:27]: sure, but it still won't work conditionally, see this modified playground based on the original example: 

https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgVwM4FMCSA7AZsbYGdARWXSgE8AaOA-Q4siygeTBmAm1TgF84uKBBBwARAAEYAQx4yAxgGsA9FHTT5MALQBHclTEBuAFDHcybJq7YUGAOpRpYMBQAUSAkWDSANgBFpGX4ALkQ6Rm9-QOkAflCkdAAPaXAfdFCYKHJ+fgBKRGM4OHluVHg9FnZOUrgAXnCGIlJ9Ng5rVHdCorgKqgBpdEpQgG0xJJSwNLEAXWouot7KADFsUOlUSks4V3zagD5wr18AmTnuw85j6NDPS6iZee6YsLBpAHN0VBHbyJPpWbgrw+AAVpI4QF84MNpvxHkVQhYACboBjoRFnbofGAAOSSMFBILBKVCOzqBwAjBiij9fAT0KDwaFKV0+LkTHxTOZLNUbGh0A4nC4oMwqO4Lr9oiEwjT7rF4nBxql0nBMtkBKyCkU1DBkFBeRgcI0mC1OudFgMhlCxsklTMqT0Wis1hstqT9uKrqdHjK-jcIp7pHC4M83QcxUDPt9-bKARGGSlIdC8kGEdhkaj0Y8sbjEvj3vSiSASbsKfaZXT40W4Myiqz2UA

Comment (TkDodo) [reply packages/react-query/src/infiniteQueryOptions.ts:27]: okay then what’s the fix please 😅

Comment (jimmycallin) [reply packages/react-query/src/infiniteQueryOptions.ts:27]: I made a PR here :) https://github.com/TanStack/query/pull/8157

Hope that clarifies things, or perhaps I'm misunderstanding something!

**Pull request merged**